docs/spec/connection-handshake/L1_2.md

```diff
@@ -0,0 +1,777 @@
+
+###### Communication assumptions
+
+A relayer intermediates communication between Alice and Bob.
+We model communication as a shared mutable state.
+Each chain exposes an API comprising read, write, as well as a queue (FIFO) capabilities.
+So there are two parts to the communication API:
+
+  1. a read/write store (holding the entire state of the chain)
+      - each module can write to this store, using function `setConnection`;
+      - and external processes -- in our case, any relayer -- can read from the store via a function `getConnection`;
+
+  2. a queue of datagrams (or messages)
+      - each module can `pop` datagrams stored in this queue;
+      - external processes (relayers) can `push` to this queue.
+
+We describe the API in the code snippet below.
+Connection handshake modules implement this interface.
+Note that we use the modifiers 'private' and 'public' to denote which API function can be invoked by external processes (e.g., relayer) and which functions are private to the module implementing this interface.
+The data types (`Identifier` and `ConnectionEnd`) are defined below.
+
+```golang
+type CommunicationLayer interface {
+    // Write a new value (a connection) to the local store, indexed by the
+    // identifier of the connection on the local party.
+    private setConnection(i Identifier, e ConnectionEnd)
+
+    // Relayer invokes this to `read` from a chains' local store.
+    public getConnection(i Identifier) // returns a `ConnectionEnd`
+    
+    // Relayer invokes this to append a datagram into a chain's 
+    // local store.
+    public push("outstandingDatagrams", g Datagram)
+
+    // A module consumes elements from its local store 
+    // "outstandingDatagrams".
+    private pop("outstandingDatagrams") // returns a Datagram
+    // The typical use case is that for each datagram, this module invokes the 
+    // corresponding handler, e.g., ConnOpenTry.
+}
+```
+
+Preconditions:
+  - the datagram `g` as well as Identifier `i` and Connection `e` (respectively for `push`, `setConnection` and `getConnection`) are non-null;
+
+Postconditions:
+  - for `setConnection`: the local store on the module implementing this function stores at position `i` the value `e`;
+  - for `push`: the `outstandingDatagrams` queue will append the Datagram `g`
+
+Error condition: none
+
+Properties of the communication layer:
+
+- **[ICS3-CommAssumption-1-RWIntegrity]** If some process returns a connection `e` from `getConnection(i)` for some module, then the module previously invoked `setConnection(i, e)`.
+
+- **[ICS3-CommAssumption-2-QueueIntegrity]** If some module `d` returns a datagram `g` from `Pop("outstandingDatagrams")`, then a process previously invoked `Push("outstandingDatagrams", g)`.
+
+Neither of the functions in this interface are blocking.
```

Comment:
This may be confusing. I meant to say that the functions guarantee to always return e.g., by timing-out or returning the adequate output (they never hang). But this would probably deserve more explanation then we want to go into spec. So I'll remove this detail.